### PR TITLE
Add flags to define CSRF cookie expiration time and to allow CSRF cookies per request

### PR DESCRIPTION
**Pull request by miguelborges99**
## Description

This change adds a flag to define a expiration time for the CSRF cookie, and adds a flag to allow CSRF cookies per request.
If flag "--cookie-csrf-per-request" is set to true, then  a part of the state nonce is added to the CSRF name, making it possible to have parallel callback requests.

When having parallel ajax requests after the session cookie has expired multiple callbacks occur at the same time, creating multiple CSRF tokens with the name making it impossible for the proxy to authenticate.
It solves https://github.com/oauth2-proxy/oauth2-proxy/issues/1451 and possibly https://github.com/oauth2-proxy/oauth2-proxy/issues/817

Part of the pull request is copied from other pull request [1504](https://github.com/oauth2-proxy/oauth2-proxy/pull/1504), made by [YoavOst](https://github.com/YoavOst).

## Motivation and Context

In face of multiple oauth2 proxy callbacks, the CSRF cookie faces several concurrency problems:

- some requests may fail due to missing CSRF token, in case a concurrent request cleans up the cookie, or
- some requests may fail with CSRF token mismatch, since CSRF cookie value was overriden by other concurrent request.

This fix seems to solve both issues https://github.com/oauth2-proxy/oauth2-proxy/issues/1451 and https://github.com/oauth2-proxy/oauth2-proxy/issues/817 which have more than 2 years.

## How Has This Been Tested?

Deployed using docker-compose with OIDC (keycloak) provider and cookie session store and solve the parallel AJAX requests issue.

## Checklist:

- [x] My change requires a change to the documentation or CHANGELOG.
- [x] I have updated the documentation/CHANGELOG accordingly.
- [x] I have created a feature (non-master) branch for my PR.

### CODE REVIEW

**Comment by mbode** [comment]
Getting this merged & released would really help us, thank you! ❤️

**Comment by tatellos** [comment]
Thank you for this fix! The above discussion does not quite capture the anxiety and stress that comes along with opening the browser in the morning and losing a random open tab every time.

The unpredictability - and even unability to assess in hindsight - of whether this tab was trivial information, or contained an important document that I planned to read and would not be reminded to read without this open tab again, is currently a very frustrating part of the keycloak/oauth2 experience.

**Comment by WolfgangMehner** [comment]
We're using this successfully in a production environment with several hundred users. It succeeds in solving the problem of the "403 Permission Denied" errors for us. We would really appreciate if this got merged, since it is central for a good user experience.

Overall we are quite happy with the solution. However. this approach can lead to quite a number of cookies if an application performs AJAX requests in the background. Each of this calls will redirect to /oauth2/start if the user is not authenticated, and a new cookie will be set. The calls will fall to redirects to the identity provider because of CORS rules. So each background call will set a new cookie. This behavior should at least be mentioned in the documentation, since the solutions to this problem might be highly specific to each oauth2-proxy setup.

@maintainters Is there anything still missing from the PR, in your opinion? 

For context: The OAuth2/OpenID-Connect module for Apache does exactly the same. During login, there are potentially multiple cookies all containing part of the state information of the login flow in their name.

**Comment by miguelborges99** [comment]
> We're using this successfully in a production environment with several hundred users. It succeeds in solving the problem of the "403 Permission Denied" errors for us. We would really appreciate if this got merged, since it is central for a good user experience.
> 
> Overall we are quite happy with the solution. However. this approach can lead to quite a number of cookies if an application performs AJAX requests in the background. Each of this calls will redirect to /oauth2/start if the user is not authenticated, and a new cookie will be set. The calls will then fail to redirect to the identity provider because of CORS rules. So each background call will set a new cookie. This behavior should at least be mentioned in the documentation, since the solutions to this problem might be highly specific to each oauth2-proxy setup.
> 
> @maintainters Is there anything still missing from the PR, in your opinion?
> 
> For context: The OAuth2/OpenID-Connect module for Apache does exactly the same. During login, there are potentially multiple cookies all containing part of the state information of the login flow in their name.

These CSRF cookies will be short lived, by default 15 minutes. But, you can reduce the time even further, using the flag --cookie-csrf-expire (e.g.: --cookie-csrf-expire=5m).

File: CHANGELOG.md [hunk 1]
```diff
@@ -3,8 +3,15 @@
 ## Release Highlights
 
 ## Important Notes
+- [#1708](https://github.com/oauth2-proxy/oauth2-proxy/pull/1708) Enable different CSRF cookies per request (@miguelborges99)
+  - Since the CSRF cookie name is now longer it could potentially break long cookie names (around 1000 characters).
+  - Having a unique CSRF cookie per request can lead to quite a number of cookies, in case an application performs a high number of parallel authentication requests. Each call will redirect to /oauth2/start, if the user is not authenticated, and a new cookie will be set. The successfully authenticated requests will have its CSRF cookies immediatly expired, however the failed ones will mantain its CSRF cookies until they expire (by default in 15 minutes). 
+  - The user may redefine the CSRF cookie expiration time using flag "--cookie-csrf-expire" (e.g. --cookie-csrf-expire=5m). By default, it is 15 minutes, but you can fine tune to your environment.
 
 ## Breaking Changes
+- [#1708](https://github.com/oauth2-proxy/oauth2-proxy/pull/1708) Enable different CSRF cookies per request (@miguelborges99)
```
Comment (JoelSpeed) [inline CHANGELOG.md:12]: Are we sure we want to call this a breaking change? I'm not sure it's breaking enough to warrant it. 99% of users won't notice this as a negative effect will they? I'd be tempted to keep to just an important note

**Review by JoelSpeed — APPROVED** [review]
Looks good, thanks for putting this together and apologies for the delay in getting to this

File: CHANGELOG.md (hunk 1 above)
Comment (miguelborges99) [reply CHANGELOG.md]: Hi Joel, I've done as you suggested.

**Review by JoelSpeed — APPROVED** [review]
LGTM, thanks!

**Comment by mazenaissa** [comment]
Thanks, @JoelSpeed for accepting @miguelborges99 Pull Request.
When will this feature be released?

**Comment by JoelSpeed** [comment]
Planning to cut a v7.4.0 release soon

**Comment by mingfang** [comment]
this is still a problem in v7.4.0

**Comment by miguelborges99** [comment]
> this is still a problem in v7.4.0

This PR solves a particular case with CSRF cookies. 
Do you want to open an issue describing your issue, why do you think it is a problem?

**Comment by mingfang** [comment]
This PR does not fix the CSRF problem.
This issue, https://github.com/oauth2-proxy/oauth2-proxy/issues/1937, is just one of many issues about the CSRF problem.
My current workaround is to downgrade to v6.1.1.

**Comment by miguelborges99** [comment]
The CSRF problem here was due to multiple authentication requests occurring at same time. 
The issue #1937 was due to misconfiguration and it was opened by william-bohannan. Later, rshiva777 used this issue and at same time created the issue #1962, which was closed. Finally, [aaktaev](https://github.com/aaktaev) also reported issues with keycloack but did not provide more input. 

You have to do some debugging, see the logs and understand why one version is working and the other not.

**Pull request merged**